src/Platforms/Echo.Platforms.AsmResolver/Emulation/Dispatch/ObjectModel/NewObj.cs

```diff
@@ -26,16 +26,23 @@ public override DispatchResult Execute(ExecutionContext context, CilInstruction
             var method = (IMethodDescriptor) instruction.Operand;
             
             //Allocate Object
-            var allocatedObject = environment.ValueFactory.AllocateStruct(method.DeclaringType.ToTypeSignature(), true);
-            ICliValue cilValueObject = new OValue(allocatedObject, false, environment.Is32Bit);
+            var type = method.DeclaringType.ToTypeSignature();
+            var newValue = type.IsValueType
+                ? environment.ValueFactory.CreateValue(type, true)
+                : environment.ValueFactory.CreateObject(type, true);
+
+            var cilValueObject = environment.CliMarshaller.ToCliValue(newValue, type);
 
             // Pop arguments.
             int argumentCount = environment.Architecture.GetStackPopCount(instruction);
             var arguments = context.ProgramState.Stack
                 .Pop(argumentCount, true)
                 .Cast<ICliValue>()
                 .ToList();
-            arguments.Insert(0, cilValueObject);
+
+            arguments.Insert(0, type.IsValueType
+                ? new OValue(cilValueObject, true, environment.Is32Bit)
```

Comment:
Does this simulate boxing behaviour automatically? Or what is the reason for the value being wrapped in an `OValue`?

Reply:
Technically it is not an `O`, but an `&`. According to the ECMA-335, `newobj` can initialize both reference types as well as value types. It first allocates the object/struct somewhere, and then calls the constructor referenced in the operand, with argument 0 (= `this`) pointing to this allocated memory.

Reply:
We just don't have support for proper `&` pointers yet, and `O` is very similar to `&`.